README.md
# SpoddyCoiner

Google Sheets Addon that gets cryptocurrency prices, market & meta data from the CoinMarketCap API...

* Supports all cryptos / fiats listed on https://coinmarketcap.com/
    * Covers all your marginal coins/tokens/alts/shitcoins/shillcoins/moonshots/deepDDhonies
* Works 100% with *FREE* CoinMarketCap API accounts
    * Caches API responses & provides some handy tools to help reduce the number of calls and keep within your rate-limits
    * Great for free-tier accounts which have the lowest rate-limits
    * https://coinmarketcap.com/api/pricing/
* Supports over 25 coin attributes
    * Latest prices and % change over 1h, 24h, 7d & 30d ranges
    * Latest market cap and 24h trading volume
    * Tokenomics
    * Latest FCAS scores and grades
    * And all sorts of other metadata
* Convert prices between any crypto / fiat combination

SpodyCoiner pulls data only, it does not push data anywhwere, at any time. And it never will. Ever. See [a relative link](POLICY.md) for further info.


## Custom Functions

These work just like any standard function in Sheets...

* `=SPODDYCOINER(coin, [attribute], [fiat])`
    * Get latest `coin` prices / data / metadata
    * Default `attribute` is `"price"`. See below for full list of supported attributes.
    * Some attribues (such as `"price"` or `"market_cap"`) will be returned in `fiat` currency
    * Default `fiat` currency can be set in the Addons -> SpoddyCoiner menu

* `=SPODDYCOINER_CONVERT(coin, amount, coin)`
    * Convert an `amount` from one `coin` (or fiat) to another `coin` or (fiat)

All prices/quotes/conversions use CoinMarketCap latest market rates.


## Installation + Usage

1. Install the SpoddyCoiner Addon (TODO instructions - pending Addon review)
3. Get your CoinMarket API Key: https://coinmarketcap.com/api/pricing/
4. Enter the API Key in the Google Sheets addon menu (Addons -> SpoddyCoiner -> CoinMarketCap API Key)
5. Tap the SpoddyCoiner function into a cell, you'll get a tooltip with more info: `=SPODDYCOINER(`
6. See the SpoddyCoiner Addon menu for further help, preferences & tools (Addons -> SpoddyCoiner)


### Examples

* `=SPODDYCOINER( "BTC" )` - latest price of BTC in your default currency
* `=SPODDYCOINER( "DOGE", "price", "GBP" )` - latest price of DOGE in Great British Pounds
* `=SPODDYCOINER( "XRP", "market_cap", "JPY" )` - XRP market capitalization in Japanese Yen
* `=SPODDYCOINER( "NANO", "fcas_grade_full" )` - The FCAS Grade description for NANO
* `=SPODDYCOINER( A1 )` - latest price of coin in cell A1

* `=SPODDYCOINER_CONVERT( "XMR", 0.0456, "BTC" )` - convert 0.0456 XMR to BTC
* `=SPODDYCOINER_CONVERT( "ADA", 12, "GBP" )` - convert 12 ADA to GBP currency
* `=SPODDYCOINER_CONVERT( "USD", 100, A1 )` - convert $100 to the coin/currency in cell A1

NB: the functions only take a single cell range at present (eg A1).
The next major release will allow full ranges as input (eg: A1:6, A1:F1)


### Coin Attributes

The `SPODDYCOINER` function supports the following `attributes`...

* `price` - latest price in fiat currency
* `price_percent_change_1h` - price change over last 1h
* `price_percent_change_24h` - price change over last 24h
* `price_percent_change_7d` - price change over last 7d
* `price_percent_change_30d` - price change over last 30d
* `market_cap` - latest market capitalization in fiat currency
* `volume_24h` - 24h trading volume
* `circulating_supply` - number of coins/tokens currently circulating
* `total_supply` - total number of coins/tokens potentially available
* `max_supply` - maximum number of coins/tokens that will ever be available (some coins do not have a max supply)
* `fcas_score` - Fundamental Crypto Asset Score (0-1000), a measure of the fundamental health of crypto projects (only the top 300-400 coins are rated)
* `fcas_grade` -  FCAS Grade (S,A,B,C,E,F)
* `fcas_grade_full` - Full FCAS Grade description (Superb,Attractive,Basic,Caution,Fragile)
* `fcas_point_change_24h` - 24h change in FCAS score
* `fcas_percent_change_24h` - 24h change in FCAS score as a percentage
* `name` - coin name
* `description` - description of the crypto, including tokenomicsa & latest market data.
* `description_short` - first sentence of description (dropping tokenomics and market data)
* `logo` - logo url
    * Tip: wrap this in the Google Sheets `IMAGE` function to show it in the cell
    * eg: `=IMAGE(SPODDYCOINER("BTC", "logo"))`
* `date_added` - date added to CoinMarketCap (effectively the date it started)
* `year_added` - year added to CoinMarketCap
* `tags` - comma seperated list of all tags
* `tags_top_5` - comma seperated list of the first 5 tags
* `url_website` - primary website for the project (if more than 1, only 1st returned)
* `url_technical_doc` - whitepaper tech document for the project (if more than 1, only 1st returned)
* `url_explorer` - blockchain explorer for the coin/token (if more than 1, only 1st returned)
* `url_source_code` - github url for the project source code (if available)


### Tools

The SpoddyCoiner Addon menu has a number of useful tools...

* #### Refresh Selected Cells
    * Refreshes the `SPODDYCOINER` functions in the selected cells
    * Super useful if you hit an API rate limit error, simply wait a minute and refresh the cell(s)
* #### Refresh All Functions
    * Much like above but does a refresh of all `SPODDYCOINER` functions across the whole spreadsheet
    * On large sheets with lots of SpoddyCoiner function calls, you probably want to avoid refreshing all functions and instead focus on selected cells
* #### Convert Cells to Values
    * Converts the selected cells to the raw value if they contain a `SPODDYCOINR` function
    * Useful for cells containing coin attributes that are not expected to change (eg: `max_supply`, `year_added` etc.)
    * The value will obviously never update again, but this will save you API calls
* #### Clear Cahce
    * Resets the API Cache - allowing you to get the latest value from the CoinMarketCap API
    * NB: functions on the spreadsheet will not immediately update, but you will be asked if you wish to refresh all functions

        
### About the API Cache

The API cache is your friend. It stops API calls being repeated unecessarily, helping to keep inside your rate-limit. 
Free CoinMarketCap API accounts have the lowest rate-limits, but you may purchase a higher tier to increase this.

* Default cache time is `1 hour` - the data you see may be up to 1 hour old
* Cache time can be changed in the Addons -> SpoddyCoiner menu, the maximum is 6 hours (21600 seconds)
* Cache can be cleared at any time in the Addons -> SpoddyCoiner menu


### Errors

* Error messages from the CoinMarketCap API are shown in the cell
    * This can be turned off in the preferences
    * `symbol` is interchangeable for `coin` in error messages
* Common errors:
    * `Exceeded rate limit`
        * CoinMarketCap API rate limits are relatively low on their free-tier accounts
        * You can purchase a higher tier to increase rate-limits
        * Rate-limits are reset every minute
        * See the Addon -> SpoddyCoiner -> Tools menu for a handy way to refresh the `SPODDYCOINER` functions
    * `Invalid value for symbol` or `symbol is not allowed to be empty`
        * The coin ticker/symbol you've entered is not correct
        * If you're using a cell reference, check it's accurate.


CHANGELOG.md
# Changelog

All notable changes to the "SpoddyCoiner Google Sheets Addon" project will be documented in this file. 
Breaking changes between versions will be highligted.

## [1.2.3] - In progress

* Improved:
    * Build tools + process


## [1.2.2] - 2021-05-29

* Improved:
    * Added new tool "refresh selected cells"
    * Added "description_short" attribute
* Fixed:
    * Cache time not set to default on first install
    * Refresh functions now include cells that have wrapped spoddycoiner functions 

## [1.2.1] - 2021-05-27

* Improved:
    * Menu layout, added tools menu
    * Added some new tools...
        * convert cells to values :)
        * refresh all functions
    * Cache handling
    * Removed user Session usage
        * was used to lookup users current locale & set their default currency
        * default currency on first install therefore becomes 'USD'
        * reduces required permissions, increases user privacy
    * Removed RestCountries API currency code lookup (now covered by CMC API)
    * Modernised codebase for improved developer workflow
        * NPM + Webpack + eslint (airbnb)
        * clasp deployment
        * rebuild using ES6 classes
* Fixed:
    * Bug in RestCountries, breaking install
    * Props bug stopping API key error from showing

## [1.1.0] - 2021-05-16

* Breaking Change:
    * "percent_change_" attributes are now prefixed with "price_", eg: "price_percent_change_7d"
* Added:
    * tags_top_5 attribute
* Fixed: 
    * Percent_change attributes now reflect they are about prices
    * Dont treat FCAS grade not found as an error
    * Make all percent change attributes compatible with Google Sheets percent number format
    * Return date_added in native Gooogle Sheets date format
    * Make all FCAS attributes empty if coin not ranked

## [1.0.4] - 2021-05-16

* Added: missing supported attributes + descriptions to the docs

## [1.0.3] - 2021-05-16

* Added: a changelog
* Fixed: incorrect property key name for display_error_messages

## [1.0.2] - 2021-05-16

* Added: lots
* Notes: initial release


src/view/Menu.js
class Menu {
    /**
     * SpoddyCoiner Addon Menu view
     *
     * @param {SpoddyCoiner} spoddycoiner   primary controller instance
     */
    constructor( spoddycoiner ) {
        /**
         * main controller class
         */
        this.SpoddyCoiner = spoddycoiner;

        /**
         * SpoddyCoiner var name in the gloabl scope
         * needed for addMenuItem functionName's
         */
        this.app = this.SpoddyCoiner.instanceName;

        /**
         * Menu item labels
         */
        this.MENU_ABOUT_LABEL = 'About';
        this.MENU_CMC_API_KEY_LABEL = 'CoinMarketCap API Key';
        this.MENU_PREFERENCES_LABEL = 'Preferences';
        this.MENU_TOOLS_LABEL = 'Tools';
        this.MENU_REFRESH_CELLS_LABEL = 'Refresh Selected Cells';
        this.MENU_REFRESH_ALL_FUNCTIONS_LABEL = 'Refresh All Functions';
        this.MENU_CLEAR_CACHE_LABEL = 'Clear Cache';
        this.MENU_CONVERT_CELL_TO_VALUE = 'Convert Cells To Values';
        this.MENU_DEFAULT_CURRENCY_LABEL = 'Default Currency:';
        this.MENU_CACHE_TIME_LABEL = 'Cache Time:';
        this.MENU_SHOW_ERRORS_LABEL = 'Show Errors:';
        this.MENU_DOCS_LABEL = 'Docs';
        this.MENU_FUNCTIONS_LABEL = 'Functions';
        this.MENU_ATTRIBUTES_LABEL = 'Attributes';

        /**
         * Dialogue headings / texts / labels
         */
        this.ABOUT_HEADING = `${this.SpoddyCoiner.ADDON_NAME} v${this.SpoddyCoiner.VERSION}`;
        this.ABOUT_TEXT = 'Handy little functions to get data from the CoinMarketCap API.\n\nCaches the response to help reduce the number of API calls and keep within your rate-limits.\n\nhttps://github.com/SpoddyCoder/SpoddyCoiner';
        this.DOCS_FUNCTIONS_HEADING = 'Custom Functions';
        this.DOCS_FUNCTIONS_TEXT = 'Simply tap the function name into a cell to get more information.\n\n=SPODDYCOINER( coin, attribute, fiat )\n\n=SPODDYCOINER_CONVERT( coin, amount, coin )\nNB: currency ticker can be used instead of coin\n';
        this.DOCS_ATTRIBUTES_HEADING = 'Coin Attributes';
        this.DOCS_ATTRIBUTES_TEXT = '';
        this.ENTER_API_KEY_PROMPT = 'Enter your API key';
        this.CURRENT_KEY_LABEL = 'Current Key :';
        this.API_KEY_UPDATED_LABEL = 'API Key Updated';
        this.API_CACHE_TIME_HEADING = 'API Cache Time';
        this.API_CACHE_KEY_PROMPT = `New cache time in seconds (max ${this.SpoddyCoiner.Model.APICache.MAX_CACHE_TIME})`;
        this.API_CACHE_TIME_UPDATED_LABEL = 'API Cache Time Updated';
        this.NEW_CACHE_TIME_LABEL = 'New cache time :';
        this.NUM_CACHE_ITEMS_LABEL = ' API calls currently cached';
        this.CLEAR_CACHE_PROMPT = 'Do you want to reset the API cache?';
        this.API_CACHE_CLEARED_LABEL = 'API cache cleared.';
        this.REFRESH_ALL_FUNCTIONS_NOTE = 'NB: this does not clear the cache.';
        this.REFRESH_ALL_FUNCTIONS_PROMPT = 'Do you want to re-run all the SPODDYCOINER functions on the active spreadsheet?';
        this.REFRESH_SELECTED_CELLS_LABEL = 'Do you want to refresh SPOODYCOINER functions in the selected cell(s)...\n\n';
        this.NEW_CURRENCY_CODE_HEADING = 'Enter new 3 letter currency code (ISO 4217)';
        this.DEFAULT_CURRENCY_UPDATED_LABEL = 'Default Currency Updated';
        this.NEW_CURRENCY_LABEL = 'New currency code :';
        this.CURRENCY_CODE_NOT_VALID_LABEL = 'Currency code was not valid!';
        this.TURN_ERRORS_OFF_PROMPT = 'Do you want to turn errors off?';
        this.TURN_ERRORS_ON_PROMPT = 'Do you want to turn errors on?';
        this.CONVERT_CELL_TO_VALUE_CELLS_LABEL = 'Do you want to convert the selected cell(s)...\n\n';
        this.CONVERT_CELL_TO_VALUE_PROMPT = '\n...to value(s)?';
        this.CONVERT_CELL_TO_VALUE_NOTE = 'This converts SPODDYCOINER functions in the selected cell(s) to the static value the cell currently contains (which means they will no longer be updated).';
        this.GENERIC_ERROR_MESSAGE = 'There was a problem, please try again.';

        /**
         * Supported attributes and their descriptions
         */
        this.SUPPORTED_ATTRIBUTES = {
            price: 'Latest price, in fiat currency',
            price_percent_change_1h: 'Price change over last 1h, as a percentage',
            price_percent_change_24h: 'Price change over last 24h, as a percentage',
            price_percent_change_7d: 'Price change over last 7d, as a percentage',
            price_percent_change_30d: 'Price change over last 30d, as a percentage',
            market_cap: 'Latest market capitalization, in fiat currency',
            volume_24h: '24h trading volume, in fiat currency',
            circulating_supply: 'Number of coins/tokens currently circulating',
            total_supply: 'Total number of coins/tokens potentially available',
            max_supply: 'Maximum number of coins/tokens that will ever be available (some coins do not have a max supply)',
            fcas_score: 'Fundamental Crypto Asset Score (0-1000), a measure of the fundamental health of crypto projects (only the top 300-400 coins are rated)',
            fcas_grade: 'FCAS Grade (S,A,B,C,E,F)',
            fcas_grade_full: 'Full FCAS Grade description (Superb,Attractive,Basic,Caution,Fragile)',
            fcas_percent_change_24h: '24h change in score',
            fcas_point_change_24h: '24h change in score, as a percentage',
            name: 'The cryptocurrency name',
            description: 'Description of the crypto, including tokenomicsa & latest market data',
            description_short: 'First sentence of description (dropping tokenomics and market data)',
            logo: 'The coin logo url (Tip: wrap this in the Google Sheets IMAGE function to show it in the cell)',
            date_added: 'Date added to CoinMarketCap (effectively the date it started)',
            year_added: 'Year added to CoinMarketCap',
            tags: 'A comma seperated list of all tags',
            tags_top_5: 'A comma seperated list of the first 5 tags',
            url_website: 'Primary website for the project (if more than 1, only 1st returned)',
            url_technical_doc: 'Whitepaper tech document for the project (if more than 1, only 1st returned)',
            url_explorer: 'Blockchain explorer for the coin/token (if more than 1, only 1st returned)',
            url_source_code: 'Project source code (github URL, if available)',
        };
    }

    /**
     * Basic menu when no authorization has been given
     */
    addNoAuthMenu() {
        // TODO: refactor so menu is only instantiated in correct scopes
        // cant make this a class level property atm
        // because getUi is not available in all scopes (eg: custom functions)
        const ui = SpreadsheetApp.getUi();
        ui.createAddonMenu()
            .addItem( this.MENU_ABOUT_LABEL, `${this.app}.View.Menu.about` )
            .addToUi();
    }

    /**
     * Full SpoddyCoiner menu
     */
    addMenu() {
        const ui = SpreadsheetApp.getUi();
        ui.createAddonMenu()
            .addItem( this.MENU_CMC_API_KEY_LABEL, `${this.app}.View.Menu.updateCMCApiKey` )
            .addSubMenu( ui.createMenu( this.MENU_PREFERENCES_LABEL )
                .addItem( `${this.MENU_DEFAULT_CURRENCY_LABEL} ${this.SpoddyCoiner.Model.GASProps.getDefaultCurrency()}`, `${this.app}.View.Menu.updateDefaultCurrency` )
                .addItem( `${this.MENU_CACHE_TIME_LABEL} ${this.SpoddyCoiner.Model.GASProps.getCacheTime( true )}`, `${this.app}.View.Menu.updateAPICacheTime` )
                .addItem( `${this.MENU_SHOW_ERRORS_LABEL}  ${this.SpoddyCoiner.Model.GASProps.getDisplayErrorMessages( true )}`, `${this.app}.View.Menu.showErrors` ) )
            .addSubMenu( ui.createMenu( this.MENU_TOOLS_LABEL )
                .addItem( this.MENU_REFRESH_CELLS_LABEL, `${this.app}.View.Menu.refreshSelectedCells` )
                .addItem( this.MENU_REFRESH_ALL_FUNCTIONS_LABEL, `${this.app}.View.Menu.refreshCustomFunctions` )
                .addItem( this.MENU_CONVERT_CELL_TO_VALUE, `${this.app}.View.Menu.convertCellsToValues` )
                .addItem( this.MENU_CLEAR_CACHE_LABEL, `${this.app}.View.Menu.clearAPICache` ) )
            .addSeparator()
            .addSubMenu( ui.createMenu( this.MENU_DOCS_LABEL )
                .addItem( this.MENU_FUNCTIONS_LABEL, `${this.app}.View.Menu.docsFunctions` )
                .addItem( this.MENU_ATTRIBUTES_LABEL, `${this.app}.View.Menu.docsAttributes` ) )
            .addItem( this.MENU_ABOUT_LABEL, `${this.app}.View.Menu.about` )
            .addToUi();
    }

    /**
     * 'CoinMarketCap API Key' menu item
     */
    updateCMCApiKey() {
        const ui = SpreadsheetApp.getUi();
        let apiKey = this.SpoddyCoiner.Model.GASProps.getAPIKey( true ); // masked
        const promptLabel = ( apiKey === '' ) ? this.ENTER_API_KEY_PROMPT : `${this.CURRENT_KEY_LABEL} ${apiKey}\n\n${this.ENTER_API_KEY_PROMPT}`;
        const result = ui.prompt(
            this.MENU_CMC_API_KEY_LABEL,
            promptLabel,
            ui.ButtonSet.OK_CANCEL,
        );
        const button = result.getSelectedButton();
        apiKey = result.getResponseText();
        if ( button === ui.Button.OK ) {
            if ( !this.SpoddyCoiner.Model.GASProps.setAPIKey( apiKey ) ) {
                ui.alert( this.GENERIC_ERROR_MESSAGE, ui.ButtonSet.OK );
                return;
            }
            this.promptRefreshAllFunctions( this.API_KEY_UPDATED_LABEL );
        }
    }

    /**
     * 'API Cache Time' menu item
     */
    updateAPICacheTime() {
        const ui = SpreadsheetApp.getUi();
        const result = ui.prompt(
            this.API_CACHE_TIME_HEADING,
            this.API_CACHE_KEY_PROMPT,
            ui.ButtonSet.OK_CANCEL,
        );
        const button = result.getSelectedButton();
        const newTime = result.getResponseText();
        if ( button === ui.Button.OK ) {
            if ( !this.SpoddyCoiner.Model.GASProps.setCacheTime( newTime ) ) {
                ui.alert( this.GENERIC_ERROR_MESSAGE, ui.ButtonSet.OK );
                return;
            }
            ui.alert(
                this.API_CACHE_TIME_UPDATED_LABEL,
                `${this.NEW_CACHE_TIME_LABEL} ${this.SpoddyCoiner.Model.GASProps.getCacheTime( true )}`,
                ui.ButtonSet.OK,
            );
        }
    }

    /**
     * 'Clear API Cache' menu item
     */
    clearAPICache() {
        const ui = SpreadsheetApp.getUi();
        const result = ui.alert(
            this.MENU_CLEAR_CACHE_LABEL,
            this.CLEAR_CACHE_PROMPT,
            ui.ButtonSet.YES_NO,
        );
        if ( result === ui.Button.YES ) {
            if ( !this.SpoddyCoiner.Model.APICache.clear() ) {
                ui.alert( this.GENERIC_ERROR_MESSAGE, ui.ButtonSet.OK );
                return;
            }
            this.promptRefreshAllFunctions( this.API_CACHE_CLEARED_LABEL );
        }
    }

    /**
     * 'Default Currency' menu item
     */
    updateDefaultCurrency() {
        const ui = SpreadsheetApp.getUi();
        const result = ui.prompt(
            `${this.MENU_DEFAULT_CURRENCY_LABEL} ${this.SpoddyCoiner.Model.GASProps.getDefaultCurrency()}`,
            this.NEW_CURRENCY_CODE_HEADING,
            ui.ButtonSet.OK_CANCEL,
        );
        const button = result.getSelectedButton();
        const newCurrencyCode = result.getResponseText();
        if ( button === ui.Button.CANCEL || button === ui.Button.CLOSE ) {
            return;
        }
        if ( button === ui.Button.OK ) {
            if ( !this.SpoddyCoiner.Model.GASProps.setDefaultCurrency( newCurrencyCode ) ) {
                ui.alert( this.CURRENCY_CODE_NOT_VALID_LABEL, ui.ButtonSet.OK );
                return;
            }
            this.promptRefreshAllFunctions(
                `${this.DEFAULT_CURRENCY_UPDATED_LABEL}`,
                `${this.NEW_CURRENCY_LABEL} ${this.SpoddyCoiner.Model.GASProps.getDefaultCurrency()}`,
            );
        }
    }

    /**
     * 'Show Errors' menu item
     */
    showErrors() {
        const ui = SpreadsheetApp.getUi();
        let prompt = this.TURN_ERRORS_ON_PROMPT;
        if ( this.SpoddyCoiner.Model.GASProps.getDisplayErrorMessages() ) {
            prompt = this.TURN_ERRORS_OFF_PROMPT;
        }
        const result = ui.alert(
            `${this.MENU_SHOW_ERRORS_LABEL}  ${this.SpoddyCoiner.Model.GASProps.getDisplayErrorMessages( true )}`,
            prompt,
            ui.ButtonSet.YES_NO,
        );
        if ( result === ui.Button.YES ) {
            this.SpoddyCoiner.Model.GASProps.toggleErrorMessages();
            this.promptRefreshAllFunctions(
                `${this.MENU_SHOW_ERRORS_LABEL}  ${this.SpoddyCoiner.Model.GASProps.getDisplayErrorMessages( true )}`,
            );
        }
    }

    /**
     * 'Functions' menu item
     */
    docsFunctions() {
        const ui = SpreadsheetApp.getUi();
        ui.alert(
            this.DOCS_FUNCTIONS_HEADING,
            this.DOCS_FUNCTIONS_TEXT,
            ui.ButtonSet.OK,
        );
    }

    /**
     * 'Supported Attributes' menu item
     */
    docsAttributes() {
        const ui = SpreadsheetApp.getUi();
        let supportedAttrs = '';
        Object.entries( this.SUPPORTED_ATTRIBUTES ).forEach( ( [k, v] ) => {
            supportedAttrs += `"${k}"\n ${v}\n\n`;
        } );
        const text = `${this.DOCS_ATTRIBUTES_TEXT}${supportedAttrs}`;
        ui.alert( this.DOCS_ATTRIBUTES_HEADING, text, ui.ButtonSet.OK );
    }

    /**
     * 'About' menu item
     */
    about() {
        const ui = SpreadsheetApp.getUi();
        ui.alert(
            this.ABOUT_HEADING,
            this.ABOUT_TEXT,
            ui.ButtonSet.OK,
        );
    }

    /**
     * 'Refresh All Functions' menu item
     */
    refreshCustomFunctions() {
        this.promptRefreshAllFunctions(
            this.MENU_REFRESH_ALL_FUNCTIONS_LABEL,
            this.REFRESH_ALL_FUNCTIONS_NOTE,
        );
    }

    /**
     * 'Refresh Selected Cells' menu item
     */
    refreshSelectedCells() {
        const ui = SpreadsheetApp.getUi();
        const result = ui.alert(
            this.MENU_REFRESH_CELLS_LABEL,
            `${this.REFRESH_ALL_FUNCTIONS_NOTE}\n\n${this.REFRESH_SELECTED_CELLS_LABEL}${this.SpoddyCoiner.View.Sheet.getActiveCells()}`,
            ui.ButtonSet.YES_NO,
        );
        if ( result === ui.Button.YES ) {
            this.SpoddyCoiner.handleConfirmRefreshSelectedCells();
            return true;
        }
        return false;
    }

    /**
     * `Convert Cells To Values` menu item
     */
    convertCellsToValues() {
        const ui = SpreadsheetApp.getUi();
        const result = ui.alert(
            this.MENU_CONVERT_CELL_TO_VALUE,
            `${this.CONVERT_CELL_TO_VALUE_NOTE}\n\n${this.CONVERT_CELL_TO_VALUE_CELLS_LABEL} ${this.SpoddyCoiner.View.Sheet.getActiveCells()}\n${this.CONVERT_CELL_TO_VALUE_PROMPT}`,
            ui.ButtonSet.YES_NO,
        );
        if ( result === ui.Button.YES ) {
            this.SpoddyCoiner.handleConfirmConvertCellsToValues();
        }
    }

    /**
     * Refresh all SPODDYCOINER functions dialogue
     *
     * @param {string} title    title text
     * @param {string} text     additional text above the update functions prompt
     * @returns {boolean}       confirmed OK
     */
    promptRefreshAllFunctions( title = '', text = '' ) {
        const ui = SpreadsheetApp.getUi();
        const confirmText = ( text !== '' ) ? `${text}\n\n${this.REFRESH_ALL_FUNCTIONS_PROMPT}` : this.REFRESH_ALL_FUNCTIONS_PROMPT;
        const result = ui.alert(
            title,
            confirmText,
            ui.ButtonSet.YES_NO,
        );
        if ( result === ui.Button.YES ) {
            this.SpoddyCoiner.handleConfirmRefreshAllFunctions();
            return true;
        }
        return false;
    }
}

module.exports = { Menu };
